### ISSUE TITLE
ENH: StringDType and lost arena allocation

### ISSUE THREAD

**Issue by egesip** [numpy/numpy#26860]
### Proposed new feature or change:

This is for discussion/feedback, and possibly to correct my understanding of the behavior of the StringDType and its arena.

Currently  the array copy constructors (`PyArray_Copy` & friends) re-use the dtype instance to create a new array.
That means a `StringDType` array copy will use the same arena as the original array.

A common numpy usage pattern (at least for me) is to create/load an array, massage it a bit with functions that may create a copy, then pass the copy to the rest of the program, discarding the original array.

The issue with this, if my understanding is correct, is that, when we deal with a `StringDType` array and all strings happen to be "allocated" in the arena, then the array copy, which might have a long lifetime, will carry in its arena all the lost chunk of arena that was used by the original array (there is, rightfully so, no "garbage collection" on the chunk of arena that was cleared in the clear_loop, as that would imply modifying all offsets of all arrays using this arena).

If the massaging had to create 1, 2 or 3 copies, then my long-lived final array consumes 2, 3 or 4 times what it really needs.

Wouldn't it be better that each array had its own arena, by having all the copy constructors create their own instances of the custom dtype, instead of re-using the original one?

Thanks a lot!

**Comment by seberg** [comment]
Thanks for the report, but what makes you say that they don't have their own arena"
```
a = np.arange(1000).astype(np.dtypes.StringDType())
a.dtype is a.copy().dtype
```
returns `False` as the copy has a new dtype with a new arena.

**Comment by egesip** [comment]
Ah you're right, I should have checked this myself in numpy 2.0 before bothering you... Sorry.

I am stuck with v1.26 (waiting for some downstream libraries to be fully compatible), I am using the experimental dtype api there, and have a custom DType with a sidecar allocation, very similarly to StringDType.
This sharing of dtype instance between copies is what I observe in this setup.
I had scanned the 2.0 code base and had not seen anything that would explain a different behavior in 2.0, obviously I overlooked something :-)
Back to reading the 2.0 source then, I'll probably need to port the "better" behavior of 2.0 into my local 1.26, if that's not a rabbit hole.
Thank you very much!
